internal/rukpak/preflights/crdupgradesafety/crdupgradesafety.go

```diff
@@ -112,9 +114,71 @@ func (p *Preflight) runPreflight(ctx context.Context, rel *release.Release) erro
 
 		err = p.validator.Validate(*oldCrd, *newCrd)
 		if err != nil {
+			err = orderKappsValidateErr(err)
 			validateErrors = append(validateErrors, fmt.Errorf("validating upgrade for CRD %q failed: %w", newCrd.Name, err))
 		}
 	}
 
 	return errors.Join(validateErrors...)
 }
+
+// orderKappsValidateErr is meant as a temporary solution to the problem
+// of randomly ordered multi-line validation error returned by kapp's validator.Validate()
+//
+// The problem is that kapp's field validations are performed in map iteration order, which is not fixed.
+// Errors from those validations are then error.Join'ed, fmt.Errorf'ed and error.Join'ed again,
+// which means original messages are available at 3rd level of nesting, and this is where we need to
+// sort them to ensure we do not enter into constant reconciliation loop because of random order of
+// failure message we ultimately set in ClusterExtension's status conditions.
+//
+// This helper attempts to do that and falls back to original unchanged error message
+// in case of any unforeseen issues which likely mean that the internals of validator.Validate
+// have changed.
+//
+// For full context see:
+// github.com/operator-framework/operator-controller/issues/1456 (original issue and comments)
+// github.com/carvel-dev/kapp/pull/1047 (PR to ensure order in upstream)
+//
+// TODO: remove this once ordering has been handled by the upstream.
+func orderKappsValidateErr(err error) error {
```

Comment:
Well written (as with the unit tests). Thank you! Qq: is it worth surfacing whether the function had to fallback on the original? My guess is that if we fallback, we may run into the continuous reconcile issue. Wondering if we should make that noisy somehow...?

Reply:
thank you @perdasilva I thought about it too and it's a tricky situation, because while the first two steps in the unwrapping logic are true to all internal validators (as of now at least), the last one (most nested error.Join) might only be valid for `CheckValidator` internal validator. Since fallback might as well be a false negative here, I'd normally log this via `Debug`, but I noticed `Debug` isn't being used anywhere in the code so ultimately thought against it.
In case of a non-empty error here and regardless of fallback, we are still logging the full reconciliation error message every time we try to reconcile and that should make spotting the random order possible, though I agree an added hint from failed fallback might definitely help, so maybe it would be ok.

WDYT?

Reply:
yeah. Ultimately, I guess, it becomes a debug artifice. Which, while it undoubtedly helps with debugging, we'd still be reconciling repeatedly. If we have exp. backoff, then maybe it's not a big issue, and it's also temporary. I was thinking if we should panic (since the library shouldn't change out from under us during runtime), but then we'd need to be very careful about testing each error condition the underlying library checks for, which might not be a good use of time atm. Let's proceed ^^ thanks, dude =D